### PR TITLE
Ocp tools compatable

### PR DESCRIPTION
**Pull request by wadeking98**
created x-cluster-migrate script to be compatable with ocp developer tools:
- created script to automate migration
- cleaned up docs

@esune  @WadeBarnes

### CODE REVIEW

File: CrossClusterDataSteps.md [hunk 1]
```diff
@@ -27,133 +25,53 @@
...
+
+It is important that you have logged in recently to both clusters from the command line. Go to the web console of each cluster and copy and paste each login command into your terminal.
+
+
+#### 2. Initialization  
+~~~
+./x-cluster-migrate init -p default
+~~~
+Follow the prompts on the screen and it should successfully initialize the project.  
+You will have to navigate to `./cross-cluster/openshift/templates/source-pvc-migrator/source-pvc-migrator-deploy.source.local.param` and `./cross-cluster/openshift/templates/target-pvc-migrator/target-pvc-migrator-deploy.target.local.param` and comment out the line that says `SRC_IMAGE_NAMESPACE` or `DST_IMAGE_NAMESPACE` and change it to the namespace of your `pvc-migrator` build.
```
Comment (WadeBarnes) [inline CrossClusterDataSteps.md]: The scripts should automate this process.  The user should not have to edit anything manually.  Examples of how to automate the update (comment, uncomment, update values) of the param files can be found here; https://github.com/BCDevOps/matomo-openshift/blob/master/openshift/manage#L229.  Examples are also in the `updateMigratorDeplymentParams` function of the `x-cluster-migrate` script.

File: cross-cluster/openshift/x-cluster-migrate [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+  _red='\033[0;31m'
+  _yellow='\033[1;33m'
+  _nc='\033[0m' # No Color
+  echo -e \\n"${_red}${_includeFile} could not be found on the path.${_nc}"
+  echo -e "${_yellow}Please ensure the openshift-developer-tools are installed on and registered on your path.${_nc}"
+  echo -e "${_yellow}https://github.com/BCDevOps/openshift-developer-tools${_nc}"
+fi
+
+
+usage () {
```
Comment (WadeBarnes) [inline cross-cluster/openshift/x-cluster-migrate]: Usage should always be at the top of the script.

File: cross-cluster/openshift/x-cluster-migrate [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+    $0 init -p default
+    $0 migrate -p target
+    $0 migrate -p source
+
+    $0 clean -p target
+    $0 clean -p source
+
+  Commands:
+  ========
+    init -p default
```
Comment (WadeBarnes) [inline cross-cluster/openshift/x-cluster-migrate]:
```suggestion
    init -p <profile name>
```

**Review by WadeBarnes — CHANGES_REQUESTED** [review]
The new `x-cluster-migrate` scripts should follow the same pattern and flow as the exiting [manage](https://github.com/BCDevOps/StorageMigration/blob/master/openshift/manage) script:

- Initialize
  - Initialize the migration for the given environments.
  
- Build
  - Deploy the BCs and the build storage migration containers in the tools projects of the source and destination projects and clusters.
  
- Migrate
  - Fully orchestrate the migration of data from the source to the destination.  _The user should only have to issue a single command, not two._

- Clean
  - Clear out the migration resources from the given environment (the one specified with `-e`) in both clusters.  This allows the migration resource to be cleared out of each deployment environment (`dev`, `test`, `prod`) separately as the migrations are complete and then finally from the `tools` environment.
  
There are a number of empty files that need to be removed from the commit history too.

File: CrossClusterDataSteps.md (hunk 1 above)
Comment (wadeking98) [reply CrossClusterDataSteps.md]: change completed

File: cross-cluster/openshift/x-cluster-migrate (hunk 2 above)
Comment (wadeking98) [reply cross-cluster/openshift/x-cluster-migrate]: change completed

**Review by esune — CHANGES_REQUESTED** [review]
Since this repository will be used by other teams as well, the param files as well as settings files specific to a use-case should not be committed to version control. The `init` command should instead create `local` parameter files that are gitignored.

File: cross-cluster/openshift/settings.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+export GIT_URI=https://github.com/BCDevOps/StorageMigration.git
+export GIT_REF=master
```
Comment (esune) [inline cross-cluster/openshift/settings.sh:2]: This should not be committed, a `settings.local.sh` should be generated instead by the `init` step, like we do for other services (e.g.: Matomo).

File: cross-cluster/openshift/templates/nsp/pvc-migrator-nsp-deploy.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        - - ext:network=any
+parameters:
+  - name: NAME
+    description: "Name of the deployment config."
+    value: "pvc-migrator-nsp"
+  - name: NAMESPACE_NAME
+    displayName: Namespace Name
+    description: The name of the namespace being deployed to..
+    required: true
+    value: a99fd4-tools
```
Comment (esune) [inline cross-cluster/openshift/templates/nsp/pvc-migrator-nsp-deploy.yaml]: We don't want project-specific parameters to be committed to the main repository: this should either be left blank, or with a default "dummy" value like `myproject`.

File: cross-cluster/openshift/x-cluster-migrate [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+  echo "export SRC_CONTEXT=${SRC_CONTEXT}" >> settings.local.sh
+
+  # printStatusMsg "Is your source cluster on OpenShift Platform 4? (yes/no):"
+  # read SRC_OCP4_RESP
+  # if [ "${SRC_OCP4_RESP}" = "yes" ]; then 
+  #   SRC_OCP4=true 
+  # else 
+  #   SRC_OCP4=false 
+  # fi
+  # echo "export SRC_OCP4=${SRC_OCP4}" >> settings.sh
```
Comment (esune) [inline cross-cluster/openshift/x-cluster-migrate]: If the commented-out parts are not necessary they should go.

File: cross-cluster/openshift/x-cluster-migrate [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+  echo "Generating param files ..."
+  genTemplateParams.sh -lf -p default
+
+  # rename files to correct profile
+  SRC_PARAM_NAME="./templates/source-pvc-migrator/source-pvc-migrator-deploy.local.param"
+  DST_PARAM_NAME="./templates/target-pvc-migrator/target-pvc-migrator-deploy.local.param"
+
+  
+
+  updateMigratorDeplymentParams ${SRC_PVC} ${DST_PVC} ${REMOTE_CLUSTER} ${REMOTE_PROJECT} ${SRC_CLUSTER} ${DST_CLUSTER} "${SRC_PARAM_NAME} ${DST_PARAM_NAME}"
```
Comment (esune) [inline cross-cluster/openshift/x-cluster-migrate]: Nit.: typo, should be `updateMigratorDeploymentParams`

File: cross-cluster/openshift/x-cluster-migrate [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+
+  printStatusMsg "Cleaning secrets, deployments, builds, etc"
+
+  # make sure we're using the destination context
+  printStatusMsg "Switching to destination cluster"
+  oc config use-context ${_context}
+
+  oc delete -l base=xc-pvc-migrator deployment,sa,rolebinding,secret
+  oc delete -l app=pvc-migrator build,buildconfig,is
+  oc delete -l app=pvc-migrator-nsp nsp
```
Comment (esune) [inline cross-cluster/openshift/x-cluster-migrate]: We should probably be namespacing these commands with `-n something`, I suspect using the context does not limit actions to a specific namespace - if the user has access to more than one on the same cluster.

File: openshift/settings.sh [hunk 9]
```diff
@@ -1,2 +1,3 @@
-export PROJECT_NAMESPACE=${PROJECT_NAMESPACE:-"bcgov"}
-export GIT_URI=${GIT_URI:-"https://github.com/BCDevOps/StorageMigration.git"}
\ No newline at end of file
+export PROJECT_NAMESPACE=${PROJECT_NAMESPACE:-"devex-von"}
+export GIT_URI=${GIT_URI:-"https://github.com/BCDevOps/StorageMigration.git"}
+export GIT_REF=${GIT_REF:-"master"}
```
Comment (esune) [inline openshift/settings.sh:3]: I am thinking this file is a leftover from before having the `manage` script and it should go. @WadeBarnes can you please confirm?

File: openshift/templates/pvc-migrator-build.param [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#=========================================================
+# OpenShift template parameters for:
+# Component: .
+# Template File: ../openshift/templates/pvc-migrator-build.yaml
+#=========================================================
+NAME=pvc-migrator
+GIT_URI=https://github.com/BCDevOps/StorageMigration.git
+GIT_REF=master
+SOURCE_CONTEXT_DIR=docker
```
Comment (esune) [inline openshift/templates/pvc-migrator-build.param]: The param file should go. The `init` function will create a `local` version to be used. Applies to all other param files as well.

**Review by esune — CHANGES_REQUESTED** [review]

**Review by WadeBarnes — CHANGES_REQUESTED** [review]
I agree with @esune's recommendations.  The `openshift/settings.sh` is required, even for the `manage` script(s).

File: openshift/settings.sh (hunk 9 above)
Comment (WadeBarnes) [reply openshift/settings.sh:3]: Required, even for the `manage` script(s).

File: cross-cluster/openshift/x-cluster-migrate (hunk 8 above)
Comment (WadeBarnes) [reply cross-cluster/openshift/x-cluster-migrate]: Agreed!

File: openshift/templates/pvc-migrator-build.param (hunk 10 above)
Comment (WadeBarnes) [reply openshift/templates/pvc-migrator-build.param]: Agreed, `.param` files should not be checked in.

File: cross-cluster/openshift/x-cluster-migrate (hunk 6 above)
Comment (wadeking98) [reply cross-cluster/openshift/x-cluster-migrate]: change completed

File: cross-cluster/openshift/x-cluster-migrate (hunk 8 above)
Comment (wadeking98) [reply cross-cluster/openshift/x-cluster-migrate]: added namespace

File: openshift/templates/pvc-migrator-build.param (hunk 10 above)
Comment (wadeking98) [reply openshift/templates/pvc-migrator-build.param]: removed param files

File: cross-cluster/openshift/settings.sh (hunk 4 above)
Comment (wadeking98) [reply cross-cluster/openshift/settings.sh:2]: I think wade said that the settings.sh files are required

Comment (wadeking98) [reply cross-cluster/openshift/settings.sh:2]: the settings.local.sh file is generated during init, but we need both settings.sh and settings.local.sh

File: cross-cluster/openshift/templates/nsp/pvc-migrator-nsp-deploy.yaml (hunk 5 above)
Comment (wadeking98) [reply cross-cluster/openshift/templates/nsp/pvc-migrator-nsp-deploy.yaml]: change completed

File: cross-cluster/openshift/x-cluster-migrate (hunk 7 above)
Comment (wadeking98) [reply cross-cluster/openshift/x-cluster-migrate]: change completed

File: cross-cluster/openshift/settings.sh (hunk 4 above)
Comment (esune) [reply cross-cluster/openshift/settings.sh:2]: You are correct about this, sorry about the confusion.

**Review by esune — COMMENTED** [review]
Looks good to me. Leaving the final review to @WadeBarnes

**Comment by esune** [comment]
I am testing this to move data cross clusters, I am getting an error when starting the `migrate` process, so I am investigating:
```
Deploying deployment configuration files ...
invalid namespace "e79518-dev/api-silver-devops-gov-bc-ca:6443/esune@github": [may not contain '/']
invalid namespace "e79518-dev/api-silver-devops-gov-bc-ca:6443/esune@github": [may not contain '/']
invalid namespace "e79518-dev/api-silver-devops-gov-bc-ca:6443/esune@github": [may not contain '/']
```

**Comment by esune** [comment]
The error I reported earlier has been since fixed, however I am incurring in a weird behaviour: running `init` the first time results in `param` files that do not have all of the user-inputted values. Running the same command again results in the files being correctly populated.

I am also getting the following error on cleanup, I suspect this is because I had to manually create the context for the environment as signing-in using the token did not create one:
```
Error from server (Forbidden): rolebindings.authorization.openshift.io is forbidden: User "esune" cannot list rolebindings.authorization.openshift.io in the namespace "nnfbch-prod": no RBAC policy matched
```

**Comment by WadeBarnes** [comment]
@esune, Check you're permissions on the namespace, you may not have permission to `delete` the role binding; https://github.com/BCDevOps/StorageMigration/pull/57/files#diff-c5bec5c49868984f9ca9c670c773e639fe0f5fd045c600bcf6e0cb3198a3cd97R341

**Comment by esune** [comment]
> 
> 
> @esune, Check you're permissions on the namespace, you may not have permission to `delete` the role binding; https://github.com/BCDevOps/StorageMigration/pull/57/files#diff-c5bec5c49868984f9ca9c670c773e639fe0f5fd045c600bcf6e0cb3198a3cd97R341

I doubt it, I am an admin 😅 

I PR'd some changes to @wadeking98 that will be incorporated into this PR, however that issue - while not a blocker - persists.

What concerns me more is the fact that a user needs to run the `init` command twice in order for the param files to update, tested from a completely clean version of the repo (no pre-existing `local` files). This seems like a glitch in how the `init` and `genParams` scripts interact, but have not had a chance to look into it further.

**Review by esune — APPROVED** [review]
Leaving the last check to @WadeBarnes - looks good to me (lus we've been actively using the migrator to move stuff over to OCP4 for a while now).

**Review by WadeBarnes — APPROVED** [review]

**Pull request merged**